cl/alerts/tasks.py

```diff
@@ -163,7 +222,7 @@ def send_docket_alert(
         return
 
     # Notify every user that's subscribed to this alert.
-    messages = make_alert_messages(
+    messages, da_webhook_recipients = make_alert_messages(
```

Comment:
We need refactoring here. This is the general stuff we do:

 - Get recipients for messages
 - Get recipients for webhooks
 - Generate emails
 - Generate webhooks
 - Send emails
 - Send webhooks

Right now we call send_docket_alerts, which calls make_alert_messages and send_docket_to_webhook. make_alert_messages, makes the messages, and gathers a list of webhook recipients, and sends messages. send_docket_to_webhook takes recipients from make_alert_messages, then generates webhooks and sends them. This is all very confusing! 

Can you see if you can simplify this so that, perhaps, it more closely mirrors the six steps above? Somebody calls something like send_alerts_and_webhooks, and then it gets recipients for each, generates content for each, and sends each?

Reply:
I've refactored these methods as suggested, I hope with these changes everything is more clear.

- Now somebody calls `send_alert_and_webhook` or `send_alerts_and_webhooks` to start the sending docket alerts and webhooks process.
- `send_alert_and_webhook` first calls `get_docket_alert_recipients` to get recipients for docket alerts.
- `get_docket_alert_recipients` returns recipients for each type of notification including the new admin's notification to inform about @recap.email users that don't exist.
- Then we call `send_recap_email_user_not_found` to send the notification to admins if there are any @recap.email users that don't exist.
- Then we compose each type of docket alert message:
  - `docket_alert_messages`: messages for users that are currently subscribed to the case
  - `re_first_time_auto_subscribe_messages`: messages for @recap.email users that receive a notification for the case and have the auto-subscribe option enabled.
  - `re_first_time_no_auto_subscribe_messages`: messages for @recap.email users that receive a notification for the case and have the auto-subscribe option disabled.
- Then we join the different types of messages and sent them at once.
- Finally, we get webhooks recipients and sent them.

Reply:
Refactor looks good. I still wonder if it could be simplified more, but I'm happy with it. I think it's a big improvement. Hopefully it still works properly. That's a risk, but it's worth taking when the code gets so complex. Thanks for sticking with it through two rounds of review.